### PR TITLE
EVAKA-4190 Email notifications about new messages

### PR DESCRIPTION
**Pull request by akheron**
#### Summary

- Notification emails about new messages are sent as async jobs, one per email
- Wrap message sending functions in a class for easier injection of `MessageNotificationEmailService`
- Remove notification service for bulletins

### CODE REVIEW

File: service/src/integrationTest/kotlin/fi/espoo/evaka/pis/service/MergeServiceIntegrationTest.kt [hunk 1]
```diff
@@ -52,6 +52,7 @@ class MergeServiceIntegrationTest : PureJdbiTest() {
 
     private val objectMapper: ObjectMapper = defaultObjectMapper()
     private val asyncJobRunnerMock: AsyncJobRunner = Mockito.mock(AsyncJobRunner::class.java)
+    private val messageService: MessageService = MessageService(null)
```
Comment (kimifoud) [inline service/src/integrationTest/kotlin/fi/espoo/evaka/pis/service/MergeServiceIntegrationTest.kt:55]: With non-nullable dependency
```kt
    private val notificationServiceMock = Mockito.mock(MessageNotificationEmailService::class.java)
    private val messageService: MessageService = MessageService(notificationServiceMock)
```

File: service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageService.kt [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    val childFirstName: String,
+    val childLastName: String,
+    val childDateOfBirth: LocalDate,
+    val receiverPersons: List<MessageReceiverPerson>
+)
+data class MessageReceiversResponse(
+    val groupId: UUID,
+    val groupName: String,
+    val receivers: List<MessageReceiver>
+)
```
Comment (kimifoud) [inline service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageService.kt:87]: I would keep these types in `Message.kt`. Only some of them are used by `MessageService`

File: service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageService.kt [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+class MessageService(
+    private val notificationEmailService: MessageNotificationEmailService?
```
Comment (kimifoud) [inline service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageService.kt:90]: non-nullable as discussed

File: service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageNotificationEmailService.kt [hunk 4]
```diff
@@ -147,7 +119,7 @@ class BulletinNotificationEmailService(
         val messagesUrl = getCitizenMessagesUrl(language)
         return when (language) {
             Language.en -> """
-                <p>You have received a new eVaka bulletin/message. Read the message here: <a href="$messagesUrl">$messagesUrl</a></p>
+                <p>You have received a new eVaka Message/message. Read the message here: <a href="$messagesUrl">$messagesUrl</a></p>
```
Comment (kimifoud) [inline service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageNotificationEmailService.kt:122]: emacs ftw?
```suggestion
                <p>You have received a new eVaka bulletin/message. Read the message here: <a href="$messagesUrl">$messagesUrl</a></p>
```

File: service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageNotificationEmailService.kt [hunk 5]
```diff
@@ -100,36 +81,27 @@ class BulletinNotificationEmailService(
...
+            tx.markNotificationAsSent(messageRecipientId)
         }
     }
 
     private fun getSubject(language: Language): String {
         val postfix = if (System.getenv("VOLTTI_ENV") == "prod") "" else " [${System.getenv("VOLTTI_ENV")}]"
 
         return when (language) {
-            Language.en -> "New bulletin in eVaka$postfix"
+            Language.en -> "New Message in eVaka$postfix"
```
Comment (kimifoud) [inline service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageNotificationEmailService.kt:104]:
```suggestion
            Language.en -> "New message in eVaka$postfix"
```

File: service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageNotificationEmailService.kt [hunk 6]
```diff
@@ -165,7 +137,7 @@ class BulletinNotificationEmailService(
         val messagesUrl = getCitizenMessagesUrl(language)
         return when (language) {
             Language.en -> """
-                You have received a new eVaka bulletin/message. Read the message here: $messagesUrl
+                You have received a new eVaka Message/message. Read the message here: $messagesUrl
```
Comment (kimifoud) [inline service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageNotificationEmailService.kt:140]:
```suggestion
                You have received a new eVaka bulletin/message. Read the message here: $messagesUrl
```

File: service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageNotificationEmailService.kt (hunk 5 above)
Comment (kimifoud) [reply service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageNotificationEmailService.kt:104]: Actually all the subject versions should be fixed now that we have messages too.

**Review by kimifoud — CHANGES_REQUESTED** [review]

File: service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageNotificationEmailService.kt (hunk 4 above)
Comment (akheron) [reply service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageNotificationEmailService.kt:122]: I actually used IDEA's search&replace, and here's the result! 😠 Should stick to Emacs.

File: service/src/integrationTest/kotlin/fi/espoo/evaka/pis/service/MergeServiceIntegrationTest.kt (hunk 1 above)
Comment (akheron) [reply service/src/integrationTest/kotlin/fi/espoo/evaka/pis/service/MergeServiceIntegrationTest.kt]: Done

File: service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageService.kt (hunk 2 above)
Comment (akheron) [reply service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageService.kt]: Moved back

File: service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageService.kt (hunk 3 above)
Comment (akheron) [reply service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageService.kt]: Done

File: service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageNotificationEmailService.kt (hunk 5 above)
Comment (akheron) [reply service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageNotificationEmailService.kt]: I kept the subjects shorter, bulletin is mentioned in the message body

**Review by kimifoud — APPROVED** [review]
LGTM 👍

**Pull request merged**